---
title: "Explore data integration and batch effects"
author: 
  - name: "Almut Luetge"
    affiliation:
       - &IMLS Institute for Molecular Life Sciences, University of Zurich, Switzerland
       - &SIB SIB Swiss Institute of Bioinformatics, University of Zurich, Switzerland
    email: "almut.luetge@uzh.ch"
  - name: Mark D Robinson
    affiliation:
      - *IMLS
      - *SIB
package: "`r BiocStyle::Githubpkg('almutlue/CellMixS')`"
output: 
    BiocStyle::html_document
bibliography: cellmixs.bib
abstract: >
  A tool set to evaluate and visualize data integration and batch effects in single cell RNA-seq data.  
vignette: >
    %\VignetteIndexEntry{Explore data integration and batch effects}
    %\VignetteEncoding{UTF-8}  
    %\VignetteEngine{knitr::rmarkdown}
editor_options: 
  chunk_output_type: console
---

```{r v1, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>",
  eval = TRUE
)
```

# Introduction

The `r BiocStyle::Githubpkg('almutlue/CellMixS')` package is a toolbox to explore and compare group effects in single cell RNA-seq data. 
It has two major aplications:  
  
* Detection of batch effects and bias in single cell RNA-seq data.    
* Evaluation and comparison of data integration (e.g. after batch effect correction).  
  
For this purpose it introduces two new metrics:  

* **Cellspecific Mixing Score (CMS)**: A test for random group mixing within k-nearest neighbouring cells.     
* **Local Density Differences (ldfDiff)**: A score describing the change in relative local cell densities by data integration or projection. 

Besides this, several explorative ploting functions enable evaluation of key integration and mixing features.  


# Installation

`r BiocStyle::Githubpkg('almutlue/CellMixS')` can be installed from github using the `r BiocStyle::CRANpkg("devtools")` package from CRAN.

```{r install, eval=FALSE}
if (!requireNamespace("devtools", quietly = TRUE))
    install.packages("devtools")
devtools::install_github("almutlue/CellMixS")
```

After installation the package can be loaded into R. 
```{r load, message=FALSE}
library(CellMixS)
```

# Getting started

## Load example data
`r BiocStyle::Githubpkg('almutlue/CellMixS')` uses the `SingleCellExperiment` class from the `r BiocStyle::Biocpkg("SingleCellExperiment")` as format for input data. 

Here we provide **sim_30**, a list of simulated single cell RNA-seq data with varying batch effect strength and unbalanced batch sizes. 

Batch effects were introduced by sampling 1%, 10%, 30% or 50% of gene expression values from a distribution with variant mean (e.g. 1% - 50% of genes were affected by a batch effect). 

All datasets consist of *3 batches*, one with _500 cells_ and the others with _150 cells_. Simulation is modified after [@Buttner2019] and described in [sim30](https://github.com/almutlue/CellMixS/blob/master/inst/script/simulate_batch_scRNAseq.Rmd).

```{r warning=FALSE}
# load required packages
suppressPackageStartupMessages({
    library(SingleCellExperiment)
    library(cowplot)
    library(limma)
})
```

```{r data}
# load sim_30 example data
load(system.file("extdata/sim30.rda", package = "CellMixS"))
names(sim_30)

sce50 <- sim_30[["batch50"]]
is(sce50)

table(sce50[["batch"]])

```

## Visualize batch effect 

Often batch effects can already be detected by eye and simple visualization (e.g. in a normal tsne or umap) depending on the strength. `r BiocStyle::Githubpkg('almutlue/CellMixS')` has different plotting functions to visualize group label and mixing scores aside without the need of using different packages. Results are `ggplot` objects and can be further customized using `r BiocStyle::CRANpkg("ggplot2")`. Other packages as `r BiocStyle::Biocpkg("scater")` provide similar plotting functions and could be used as well.

```{r vis batch 50 }
#visualize batch distribution in sce50
visGroup(sce50, group = "batch")

```

```{r vis batch all, fig.wide=TRUE}
#visualize batch distribution in other elements of sim_30 
batch_names <- c("batch1", "batch10", "batch30")
  
vis_batch <- lapply(batch_names, function(name){
    sce <- sim_30[[name]]
    visGroup(sce, "batch") + ggtitle(paste0("sim30_", name))
})

plot_grid(plotlist = vis_batch, ncol = 3)
```



# Quantify batch effects

## Cellspecific Mixing scores

Not all batch effects or group differences get obvious using visualization. Also in single cell experiments celltypes and cells can be affected in differently by experimental coditions causing batch effects (e.g. some cells are more robust for storing conditions than others).  

Furthermore the range of methods for data integration and batch effect removal gives rise to the question which method performs best on which data and thereby to quatify batch effects.  

The **cellspecific mixing score** `cms` tests for each cell the hypothesis that batch-specific distance distributions towards it's k-nearest neighbouring (knn) cells are derived from the same unspecified underlying distribution using the Anderson-Darling test [@Scholz1987]. Results from the `cms` function are two scores *cms* and *cms_smooth*, which is the weighted mean within each cell's neighbourhood. 

```{r cms}
#call cell-specific mixing score for sce50
cms50 <- cms(sce50, k = 80, group = "batch")
head(cms50)

#call cell-specific mixing score for all
cms_sim30 <- lapply(batch_names, function(name){
    sce <- sim_30[[name]]
    cms(sce, k = 80, group = "batch")
})
names(cms_sim30) <- batch_names

#append cms50
cms_sim30[["batch50"]] <- cms50

```

## Cms parameter setting

The most important parameter to set to calculate `cms` is `k `, the number of *knn* cells to use in the Anderson-Darling test. The optimal choice depend on the application, as with a small `k` focus is on local mixing, while with a large `k` mixing with regard to more global structures is evaluated. If the overall dataset structure is very heterogen with large differences in the number of cells per celltypes, it might be useful to adapt the number of *knn*. This can be done by setting the k_min parameter to the minimum number of *knn* cells to include. Before performing the hypothesis test `cms` will look for local minima in the *overall distance distribution* of it's *knn* cells. Only cells within a distance smaller than the first local minimum are then included. This can e.g. ensure that only cells from the same celltype cluster are included without relying on previous clustering algorithms. 
  
Another important parameter is the subspace to use to calculate cell distances. This can be set using the `dim_red` parameter. In default *PCA* subspace will be used and calculated if not present. Some *data integration methods* provide embeddings of a *common subspace* instead of "corrected counts". `Cms` scores can be calculated within these by defining them in `dim_red` (see \@ref(di1))


## Visualize cms

An overall summary of `cms` can be visualized as histogram. As `cms` score are *p.values* from hypothesis testing without any batch effect the p.value histogram should be flat. An increased number of very small p.values indicate the presence of a batch-specific bias within data.

```{r hist, fig.wide= TRUE}
#pval hist of cms50
visHist(cms50)

#pval hist cms_sim30
#combine cms results in one matrix
batch_names <- names(sim_30)
cms_mat <- do.call(cbind, lapply(batch_names, function(name){
  cms_res <- cms_sim30[[name]][, "cms"]
}))
colnames(cms_mat) <- batch_names

visHist(cms_mat, ncol = 4)

```


Results of `cms` can be visualized cellspecific and alongside any metadata. 
```{r single plots, fig.wide= TRUE}
#cms only cms10
cms10 <- cms_sim30[["batch10"]]
sce10 <- sim_30[["batch10"]]
cms_sce10 <- visCms(cms_res = cms10, sce = sce10, cms_var = "cms_smooth")

#group only cms10
group_sce10 <- visGroup(sce10, group = "batch")

plot_grid(cms_sce10, group_sce10, ncol = 2)

```


```{r overview}
#add random celltype assignments as new metadata
sce10[["celltype"]] <- sample(c("CD4+", "CD8+", "CD3", "CD19"), ncol(sce10), replace = TRUE)

visOverview(cms10, sce10, "batch", other_Var = "celltype")
```

Systematic differences (e.g. celltype differences) can be further explored using `compareCluster`. Here we do not expect any systematic difference as celltypes were randomly assigned.
```{r compare cluster, fig.small= TRUE}
compareCluster(cms10, "celltype", sce = sce10)
```


# Evaluate data integration 

## Mixing after data integration {#di1}

To remove or avoid batch effects when integrating different single cell RNAseq datasets a range of methods can be used. `cms` function can be used to evaluate their performance. Some of them (e.g. `fastMNN` from the `r BiocStyle::Biocpkg("scran")` package) provide a "common subspace" with integrated embeddings. Other methods like `r BiocStyle::Biocpkg("limma")` give "batch-corrected data" as results. Both work as input for `cms`.
```{r batch correction methods}
#MNN - embeddings are stored in the reducedDims slot of sce
reducedDimNames(sce10)
cms_mnn <- cms(sce10, k = 80, group = "batch", dim_red = "MNN")

# run limma
limma_corrected <- removeBatchEffect(counts(sce10), batch = sce10$batch)
#add corrected counts to sce
assay(sce10, "lim_corrected") <- limma_corrected 

#run cms
cms_lim <- cms(sce10, k = 80, group = "batch", assay_name = "lim_corrected")

```


## Compare data integration methods {#di2}

To compare different methods summary plots from `compareIntegration` and p-value histograms from `visHist` can be used. Local pattern within single methods can be explored as described above.
```{r batch correction methods vis}
#combine results
cms_res <- data.frame("mnn" = cms_mnn[,"cms"], "limma" = cms_lim[,"cms"], "unaligned" = cms10[,"cms"])

# plot results
compareIntegration(cms_res, violin = TRUE)

# As pvalue histograms
visHist(cms_res, ncol = 3)
```

Here both methods `r BiocStyle::Biocpkg("limma")` and `fastMNN` from the `r BiocStyle::Biocpkg("scran")` package show a flat p.value distribution. So cells are equally mixed after batch effect removal.

## Remaining batch-specific structure - ldfDiff

Besides successful batch "mixing" data integration should also preserve the data's internal structure and variability without adding new sources of variability or removing underlying structures. Especially for methods that result in "corrected counts" it is important to understand to much the dataset internal structures are preserved.  
 
`ldfDiff` calculates the differences between each cell's **local density factor** before and after data integration [@Latecki2007]. The local density factor is a relative meassure of the cell density around a cell compared to the densities within it's neighbourhood. Local density factors are calculated on the same set of k cells from the cell's kNN before integration. In an optimal case relative densities (according to the same set of cells) should not change by integration and the `ldfDiff` score should be close to 0. In general the overall distribution of `ldfDiff` should be centered around 0 without long tailings.

```{r ldfDiff}
#Prepare input 
# list with single SingleCellExperiment objects 
sce_pre_list <- list("batch1" = sce10[,sce10$batch == "1"], "batch2" = sce10[,sce10$batch == "2"], "batch3" = sce10[,sce10$batch == "3"])

ldf_mnn <- ldfDiff(sce_pre_list, sce_combined = sce10, group = "batch", k = 75, dim_red = "PCA", dim_combined = "MNN", assay_pre = "counts", n_dim = 20)

ldf_lim <- ldfDiff(sce_pre_list, sce_combined = sce10, group = "batch", k = 75, dim_red = "PCA", dim_combined = "PCA", assay_pre = "counts", assay_combined = "lim_corrected",  n_dim = 20)

```


## Visualize ldfDiff

Results from `ldfDiff` can be visualized in a similar way as results from `cms`.
```{r vis ldfDiff}
#ldfDiff scores in reduced dim representation
lim_plot <- visCms(ldf_lim, sce10, cms_var = "diff", dim_red = "TSNE") + ggtitle("ldfDiff - limma")

mnn_plot <- visCms(ldf_mnn, sce10, cms_var = "diff", dim_red = "MNN") + ggtitle("ldfDiff - mnn")

plot_grid(lim_plot, mnn_plot, ncol = 2)

#ldfDiff score summarized
# combine results
ldf_res <- data.frame("mnn" = ldf_mnn[,"diff"], "limma" = ldf_lim[,"diff"])

compareIntegration(ldf_res) + labs(title="Summarized ldfDiff",x="method", y = "diff")

```

`ldfDiff` shows a clear difference between both methods. While `r BiocStyle::Biocpkg("limma")` is able to preserve the batch internal structure within batches, `fastMNN` clearly changes it. Even if batches are well mixed (see \@ref(di2)), `fastMNN`does not work for batch effect removal on these simulated data.  

# Session info
```{r session info}
sessionInfo()
```

  
# References  
 
 
